contrib/mesos/pkg/minion/tasks/task.go

```diff
@@ -0,0 +1,369 @@
+	if len(t.env) > 0 {
+		cmd.Env = t.env
+	}
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Setpgid:   true,
+		Pdeathsig: syscall.SIGKILL, // see cmdProcess.Kill
+	}
+
+	// last min check for shouldQuit here
+	select {
+	case <-t.shouldQuit:
+		t.tryError(fmt.Errorf("task execution canceled, aborting process launch"))
+		return taskShouldRestart
+	default:
+	}
+
+	if err := cmd.Start(); err != nil {
+		t.tryError(fmt.Errorf("failed to start task process %q: %v", t.bin, err))
+		return taskShouldRestart
+	}
+	log.Infoln("task started", t.name)
+	t.cmd = &cmdProcess{delegate: cmd}
+	return taskRunning
+}
+
+type exitError interface {
+	error
+
+	// see os.ProcessState.Sys: returned value can be converted to something like syscall.WaitStatus
+	Sys() interface{}
+}
+
+func taskRunning(t *Task) taskStateFn {
+	waiter := t.cmd.Wait
+	type waitResult struct {
+		completed *TaskCompletion
+		err       error
+	}
+	var sendOnce sync.Once
+	trySend := func(wr *waitResult) {
+		// guarded with once because we're only allowed to send a single "result" for each
+		// process termination. for example, if Kill() results in an error because Wait()
+		// already completed we only want to return a single result for the process.
+		sendOnce.Do(func() {
+			if wr.err != nil {
+				t.tryError(wr.err)
+			} else if wr.completed != nil {
+				t.tryComplete(wr.completed)
+			}
+		})
+	}
+	// listen for normal process completion in a goroutine; don't block because we need to listen for shouldQuit
+	waitCh := make(chan *waitResult, 1)
+	go func() {
+		wr := &waitResult{}
+		defer func() {
+			waitCh <- wr
+			close(waitCh)
+		}()
+
+		if err := waiter(); err != nil {
+			if exitError, ok := err.(exitError); ok {
+				if waitStatus, ok := exitError.Sys().(syscall.WaitStatus); ok {
+					wr.completed = &TaskCompletion{
+						name: t.name,
+						code: waitStatus.ExitStatus(),
+					}
+					return
+				}
+			}
+			wr.err = fmt.Errorf("task wait ended strangely for %q: %v", t.bin, err)
+		} else {
+			wr.completed = &TaskCompletion{name: t.name}
+		}
+	}()
+
+	// wait for the process to complete, or else for a "quit" signal on the task (at which point we'll attempt to kill manually)
+	select {
+	case <-t.shouldQuit:
+		// check for tie
+		select {
+		case wr := <-waitCh:
+			// we got a signal to quit, but we're already finished; attempt best effort delvery
+			trySend(wr)
+		default:
+			// Wait() has not exited yet, kill the process
+			log.Infof("killing %s : %s", t.name, t.bin)
+			pid, err := t.cmd.Kill()
+			if err != nil {
+				trySend(&waitResult{err: fmt.Errorf("failed to kill process: %q pid %d: %v", t.bin, pid, err)})
+			}
+			// else, Wait() should complete and send a completion event
+		}
+	case wr := <-waitCh:
+		// task has completed before we were told to quit, pass along completion and error information
+		trySend(wr)
+	}
+	return taskShouldRestart
+}
+
+// forwardUntil forwards task process completion status and errors to the given output
```

Comment:
I'd rather keep as-is: it changes task state and in this case I'd rather do that in a func that has a task receiver vs. a package-level func.